### PR TITLE
MRG+1: Add vector output for LCMV

### LINKED ISSUE #5067: ENH: LCMV and vector source solution
We recently added support for VectorSourceEstimate in MNE solutions. It seems like this could be added to LCMV, too, by changing `_apply_lcmv`.

@britta-wstnr do you see any reason not to allow this?

### PR DESCRIPTION
**Pull request by larsoner**
I wanted to see how hard it would be to do #5067, and based on the nice factoring of code done by @britta-wstnr and @wmvanvliet it was only +59/-21 with a test and documentation!

It's still up for discussion if this makes sense, but I think that any time we allow `pick_ori=None` to mean "pool orientations by norm", we should also (eventually) allow `pick_ori='vector'`.

Closes #5067.

### CODE REVIEW

**Comment by wmvanvliet** [comment]
Nice! But we should test this a bit with a simple simulation: put a single dipole source with known orientation and see if the VectorSourceEstimate approximates this orientation. We can adapt the DICS tutorial for this purpose.

For the LCMV beamformer, the current code looks good to me. For the DICS beamformer, we'll need a slightly different approach because power is always positive, so when `pick_ori=None`, all XYZ components are always positive, so instead, the direction should be the direction of max-power.

**Comment by larsoner** [comment]
>  But we should test this a bit with a simple simulation: put a single dipole source with known orientation and see if the VectorSourceEstimate approximates this orientation

Isn't this really a more general question for the correctness of our LCMV implementation?

I ask because I've implemented a variant of this and it doesn't seem to give the correct orientation (see latest commit). Maybe I am doing something wrong, though.

**Comment by britta-wstnr** [comment]
Sorry that I am so late to the game!
So is the idea here to simply return all 3 orientations, @larsoner ? If so, then I think this is a great idea and should be the default with vector beamformers.
The naming ``pick_ori = "vector"``(again) might be misleading, since combining the 3 orientations, which is the current behaviour, is still a vector beamformer. 
The question to me is, whether things like combining the orientations could move to a separate function that can be applied _after_ the actual beamforming business - or could be left to the user (same for absolute values, power time series and the like) - and then the naming would fit, too.

I will look into the orientation problem you describe above ASAP, I assume this refers to the test?

**Comment by larsoner** [comment]
> The naming pick_ori = "vector"(again) might be misleading, since combining the 3 orientations, which is the current behaviour, is still a vector beamformer.

We should probably solve this via proper documentation saying `pick_ori='vector'` and `pick_ori=None` both compute a vector beamformer, which is specified is the estimate *orientation* that is returned (one for each direction or pooled to a single value).

> I will look into the orientation problem you describe above ASAP, I assume this refers to the test?

Yes, I tried for a while to get it to give something I thought should be true but couldn't. Could be that the test isn't very good, though.

**Comment by wmvanvliet** [comment]
A beamformer needs both good quality data and noise in order to estimate a proper cov matrix. Simulations are not that straightforward, as you noticed with the DICS tutorial.

**Comment by larsoner** [comment]
> A beamformer needs both good quality data and noise in order to estimate a proper cov matrix

I see two problems that need to be solved:

1. Estimation problem: get good estimates of `data_cov` and `noise_cov`
2. Beamforming problem: combine `data_cov`, `noise_cov`, `fwd` and `reg`

It sounds like you are saying that (1) could be the problem. But in #5069 there should be no cov estimation problem -- the data are simulated, so ground truth is available for the underlying signal (noise-free) data covariance and AWGN noise covariance. Assuming I did all the math correctly (e.g., adding the covariances noise and data covariances to get the "data_cov" you would estimate in any real application) I don't think it can be an estimation problem. I think it has to be a problem with (most likely) "I did all the math correctly" or (much less likely) the beamforming code (2).

File: mne/beamformer/tests/test_lcmv.py [hunk 1]
```diff
@@ -99,6 +99,49 @@ def _get_data(tmin=-0.1, tmax=0.15, all_forward=True, epochs=True,
...
+    noise_cov = mne.make_ad_hoc_cov(info)
+    noise_cov['data'] = np.diag(noise_cov['data'])
+    evoked = mne.EvokedArray(forward_sim['sol']['data'], info)
+    with pytest.raises(ValueError, match='pick_ori must be one of'):
+        make_lcmv(evoked.info, forward, data_cov, 0.05, noise_cov,
+                  pick_ori='bad')
+    angles = list()
+    for ti in range(n_vertices):
+        data_cov['data'] = (np.outer(evoked.data[:, ti], evoked.data[:, ti]) +
+                            noise_cov['data'])
```
Comment (britta-wstnr) [inline mne/beamformer/tests/test_lcmv.py]: Hi @larsoner ,
could the problem lie here? The noise covariance matrix and data covariance matrix have quite a different order of magnitude, which means that this does not create a diagonal matrix and thus ``data_cov`` is highly rank-deficient (rank of 1), which then creates problems when taking the inverse in ``make_lcmv``.

Comment (larsoner) [reply mne/beamformer/tests/test_lcmv.py]: I tried changing the relative orders of magnitude by changing the simulated amplitude above and it had little effect

File: mne/beamformer/tests/test_lcmv.py [hunk 2]
```diff
@@ -99,6 +99,49 @@ def _get_data(tmin=-0.1, tmax=0.15, all_forward=True, epochs=True,
...
+                                   noise_cov, pick_ori='vector')
+        stc = apply_lcmv(evoked, filters)
+        assert isinstance(stc, mne.SourceEstimate)
+        stc_vector = apply_lcmv(evoked, filters_vector)
+        assert isinstance(stc_vector, mne.VectorSourceEstimate)
+        assert_allclose(stc.data, stc_vector.magnitude().data)
+        # Check the orientation
+        sim_ori = forward_sim['source_nn'][ti]
+        lcmv_ori = stc_vector.data[ti, :, ti]
+        lcmv_ori /= np.linalg.norm(lcmv_ori)
```
Comment (britta-wstnr) [inline mne/beamformer/tests/test_lcmv.py]: I've been pondering (and playing around) for a while, and have some questions:
1) at least in ``forward`` and ``forward_sim`` the ``source_nn`` are not ordered the same, does this matter here?
2) why is the time vector picked with ``ti`` as well (doesn't matter I guess )?

File: mne/beamformer/tests/test_lcmv.py [hunk 3]
```diff
@@ -99,6 +99,49 @@ def _get_data(tmin=-0.1, tmax=0.15, all_forward=True, epochs=True,
...
+        angles.append(np.rad2deg(np.arccos(np.dot(lcmv_ori, sim_ori))))
+    assert np.mean(angles) < 45
+
```
Comment (britta-wstnr) [inline mne/beamformer/tests/test_lcmv.py]: Is this based on the idea that the orientations should be driven by the leadfield?

Comment (larsoner) [reply mne/beamformer/tests/test_lcmv.py]: The data are simulated as a dipole oriented normal to the cortex. This tests to see if the beamformer recovers this orientation.

File: mne/beamformer/tests/test_lcmv.py (hunk 2 above)
Comment (larsoner) [reply mne/beamformer/tests/test_lcmv.py]: The ordering could matter, I need to check.

The data are simulated as a `np.eye(n_src)` in source space. So at the output stage, when you pick a "time point" you pick a vertex. Or at least that's what I meant to do.

**Comment by larsoner** [comment]
Okay @britta-wstnr I enhanced the test to show that `fit_dipole` recovers the orientations to some accuracy, and ensure that the data covariance is not rank-deficient. It still does not pass.

In fact it passes in the opposite direction -- it tends toward 130 deg error (whereas the dipole fit tends toward 50 deg). Is it possible that the beamformer sometimes (or always) finds the reverse orientation? In other words, is it possible that it could orient inward rather than outward (have a sign ambiguity)?

**Comment by larsoner** [comment]
Turns out I had an indexing problem. It has been fixed. Now the average error for MNE is < 35 deg, for dipole fitting it's about 55 digrees, and beamformer about 85 degrees.

I'm beginning to doubt this tells us much about whether or not `pick_ori='vector'` mode works, though. Maybe this should test should be separated, since it does not seem to test this PR's functionality so much as the beamformer itself? WDYT @wmvanvliet ?

**Comment by larsoner** [comment]
... actually the angles are:
```
[85.50261515334924, 90.49324153124446, 69.56934482628559, 88.74058771535627, 90.34196568876331, 89.03438883260176]
```
This looks too consistently close to 90, so probably something systematic. Maybe it has to do with rotation to/from surf ori, or an ordering issue. So it should probably be fixed :(

**Comment by wmvanvliet** [comment]
It's true that this is also very much a correctness test for the LCMV/DICS beamformers. Nevertheless, having `pick_ori='vector'` is only useful if it gives reliable results. This PR may be a good place to do our simulations and collaborate on this issue.

I actually implemented a `pick_ori='vector'` in my original DICS beamformer code in the `conpy` package. I'll see if I can dig up my simulation code for it...

**Comment by larsoner** [comment]
> Nevertheless, having pick_ori='vector' is only useful if it gives reliable results.

This seems to work:
```
import mne

data_path = mne.datasets.sample.data_path()
subjects_dir = data_path + '/subjects'
raw = mne.io.read_raw_fif(data_path + '/MEG/sample/sample_audvis_raw.fif')
events = mne.find_events(raw, 'STI 014')
raw.load_data().pick_types()
epochs = mne.Epochs(raw, events, event_id=1)
noise_cov = mne.compute_covariance(epochs, tmax=0.)
data_cov = mne.compute_covariance(epochs, tmin=0.)
evoked = epochs.average()
forward = mne.read_forward_solution(
    data_path + '/MEG/sample/sample_audvis-meg-oct-6-fwd.fif')

inv = mne.minimum_norm.make_inverse_operator(
    evoked.info, forward, noise_cov, loose=1.)
stc_mne = mne.minimum_norm.apply_inverse(
    evoked, inv, 1. / 9., 'dSPM', pick_ori='vector')
_, peak_time = stc_mne.magnitude().get_peak(hemi='lh')
kwargs = dict(initial_time=peak_time, hemi='lh', subjects_dir=subjects_dir)
brain_mne = stc_mne.plot(figure=1, time_label='dSPM t=%0.3f', **kwargs)

forward = mne.convert_forward_solution(forward, surf_ori=True)
filters = mne.beamformer.make_lcmv(
    evoked.info, forward, data_cov, 0.05, noise_cov, pick_ori='vector')
stc_lcmv = mne.beamformer.apply_lcmv(evoked, filters)
brain_lcmv = stc_lcmv.plot(figure=2, scale_factor=1e1,
                           time_label='LCMV t=%0.3f', **kwargs)
```
dSPM:
screenshot from 2018-04-10 13-25-41

LCMV:
screenshot from 2018-04-10 13-25-43

> This PR may be a good place to do our simulations and collaborate on this issue.

If we are convinced by this example and the existing tests (other than the orientation one) that the current code works, then I don't want to leave this lingering for too long, though. Otherwise we risk leaving this functionality out because we can't track down the bugs in my test (or some corner case in the LCMV code). Would you be comfortable merging in a couple (more) weeks if we haven't tracked it down by then? If not, what test other than this orientation one would do it for you?

**Comment by wmvanvliet** [comment]
that looks pretty convincing to me!

**Comment by larsoner** [comment]
Okay it looks like LCMV does not quite make the space sufficiently smooth, so looking at the ori at the simulation vertex is insufficient. See [this gist](https://gist.github.com/larsoner/e310931745fa29ebabd834e34cf8e169) where I adapted the test into a script so people could look if they want.

LCMV (it's hard to tell but you can't see any vector coming out of the source point, so it's small amplitude there):

screenshot from 2018-04-10 14-15-52

MNE:

screenshot from 2018-04-10 14-15-53

I would say this is due to the sparse oct4 source space, but the same holds when using oct6.

I modified the test to pool the orientation estimate across other nearby vertices (within 2cm) and now it should pass.

Happy now @wmvanvliet ? :)

**Comment by wmvanvliet** [comment]
I'm very happy! How long does the test take to run? +1 for merge if test is fast enough. Cool to see more pretty `VectorSourceEstimate` results.

**Comment by larsoner** [comment]
~3 sec locally. On Travis I don't know but it's under 10 sec because it doesn't appear on the length-20 "bad" list (the fastest of which is 10 sec). One of the others is this, though, for future reference:
```
15.57s call     mne/beamformer/tests/test_lcmv.py::test_lcmv
```

**Comment by agramfort** [comment]
@wmvanvliet merge if you're happy

**Comment by wmvanvliet** [comment]
Thanks @larsoner !

**Pull request merged**